torch/csrc/jit/script/compiler.cpp

```diff
@@ -629,8 +631,9 @@ struct to_ir {
       const SugaredValuePtr& self) {
     auto params_begin = decl.params().begin();
     auto params_end = decl.params().end();
-    if (self)
+    if (self && !dynamic_cast<UserTypeValue*>(self.get())) {
```

Comment:
Thought about it more. I don't think there should be a UserDataValue sugared value at all. Assignment and attribute lookup need to work on plain SimpleValues because these objects are first class. The __init__ handling can be special cased into that logic (and it also enables us to do more strict checks, like is it unconditionally assigning or not).